### ISSUE TITLE
Kolmogorov model

### ISSUE THREAD

**Issue by jmeyers314** [rmjarvis/Piff#38]
Add a Kolmogorov model, which I think will look a lot like a Gaussian model in implementation.  Combined with a good interpolator, this could be useful for modeling the atmospheric part of a PSF.

Before I start, any recommendations for the `.fit` method?  `import lmfit`?  Rescale HSM result?  ...?

**Comment by rmjarvis** [comment]
There are a couple of things that the Gaussian model doesn't do that @gbernstein added with the PixelGrid model.  

1. We'd like to have the option of fixing the center (nominally the centroid).  Especially for something like LSST, the location of the star will be much better measured from the stack fit of a star than from any individual exposure.  But more importantly, this allows for the possibility of including small centroid shifts that aren't captured by the WCS as part of the PSF.  If these shifts are due to atmospheric deflection, it probably makes sense to keep them as part of the PSF.  So the probable procedure will be to determine the mean center from a stack and then fix that for the individual PSF estimates, letting the PSF model itself include some centroid offset.

2. The `reflux` method is supposed to take an image of a star along with a model of the star at that location and just determine the flux and center value (if the latter is not being fixed).  The use case here is to compare models of the PSF to observed stars.  Either ones that were used to do the fit or possibly test stars that were reserved from the fit.  But to do that properly, you need to rescale it by the right flux.  Currently the implementation of this in Gaussian is lame -- it just uses the sum of the pixels in the array as the flux, which is wrong if there are any masked pixels.

3. Finally, Gary implemented a `chisq` method for PixelGrid which calculates the Taylor expansion of chisq with respect to the model parameters.  This allows you to defer the actual maximization until you have all the star measurements and can fit them all at once along with the interpolation.  I don't think this one is so important to implement for Kolmogorov, but it was important for PixelGrid when some of the stars have missing pixels.  There can be some parts of the model that are literally impossible to constrain from any single star image, so you kind of have to do this more sophisticated procedure.  So feel free to pass on this one, but if you feel ambitious and want to try to understand how that method is supposed to work from following through the PIxelGrid version, then feel free to include this here too.

And if you feel really ambitious, you could consider improving the Gaussian model to include these features as well.  :)

**Comment by jmeyers314** [comment]
Sounds good.  `reflux` doesn't sound too tricky.

I'm not so sure about `chisq` though.  Is the optimization delay still possible if `chisq` isn't linear or quadratic in the parameters?

**Comment by gbernstein** [comment]
We can take a quadratic approximation to the chisq and then iterate that (essentially Newton's method of optimization).  I'll mention that this delayed-optimization approach is also useful when individual stars are undersampled, as well as when there are defective pixels.  This use case is likely to arise often, maybe less in DES than other experiments though.

**Comment by jmeyers314** [comment]
>We can take a quadratic approximation to the chisq and then iterate that

Okay.  Thanks.  That makes sense.  

I have a question about alpha/beta/chisq.  In the `StarFit` docstring, it's mentioned that

```
The params and alpha,beta,chisq are assumed to be marginalized over flux (and over center,
if it is free to vary).
```

I'm not sure I understand this.  Does this mean, for example:

<img width="694" alt="latex-image-1" src="https://cloud.githubusercontent.com/assets/3650485/18495509/735ab9fc-79d3-11e6-8583-355cddb0cbf9.png">

and similar for alpha, beta?  (with uniform priors on x, y, f, I guess?)

Or something different...  In my head, it seems like calculating chisq derivatives while minimizing w.r.t. x, y, f may also be useful.  But I haven't tried to write it all out yet.

**Comment by jmeyers314** [comment]
I'm getting blocked by offsets/centers/origins.  

How critical is the `star.image.setOrigin(0,0)` in `test_pixel.py:make_gaussian_data`?  Similarly the `use_true_center=False` kwarg to `drawImage()`?  Are these to satisfy some assumptions of `Star`?  Or can I create test data without them?

**Comment by rmjarvis** [comment]
I don't think either of those are things that you need to incorporate into your tests.  Gary had a version of make_gaussian_data that didn't use GalSim at all and did things by hand, since he wasn't able to install GalSim on his El Capitan MacBook.  But once he passed it off to me, I converted it to the equivalent functionality using GalSim.  

So these bits are just to have GalSim do the same thing as his original code that he was writing the tests to.  For your tests, you can generate the images of Kolmogorov profiles without them.

**Comment by jmeyers314** [comment]
Okay.  That helps.  

Now, just for absolute clarity, let's suppose I have a profile `prof` with `prof.centroid() = galsim.PositionD(x=0.0, y=0.0)`, and I want to draw it with world coordinate offset `(u, v)` (b/c it's a PSF with a floating center, say) onto `star.data.image` (which can have a nontrivial wcs).  Then I would do: 

```python
prof.shift(u, v).drawImage( image, offset=(star.image_pos - image.trueCenter()) )
```

**Comment by rmjarvis** [comment]
By world coordinate offset, you mean star.fit.center, right?  So I think the command would be
```
prof.shift(star.fit.center).drawImage(image, offset=(star.image_pos - image.trueCenter()) )
```
At least I think this would be correct.  :)

**Comment by jmeyers314** [comment]
I guess that depends on `_force_model_center`, but yeah.  Actually, I had in my head that `star.center` was in image coords, so thanks for pointing out that it's actually in world coords.

**Comment by rmjarvis** [comment]
> Actually, I had in my head that star.center was in image coords, so thanks for pointing out that it's actually in world coords.

Your comment made me go check to confirm, and yes, `center` is indeed in world coordinates.  There are even helper functions `offset_to_center` and `center_to_offset` in the `Star` class to help convert between the two coordinate systems using the local Jacobian of the wcs.

**Comment by jmeyers314** [comment]
What's the right way to convert HSM mom.moments_centroid into world coordinates?

Right now, I'm trying
```python
center = image.wcs.toWorld(mom.moments_centroid - image_pos)
```
which seems to work for stars/images I construct manually (similar to what's in `test_simple.py:test_Gaussian()`.  However, if I use `star.makeTarget()`, then the right line seems to be 
```python
center = image.wcs.toWorld(mom.moments_centroid - (image.trueCenter() - image_pos))
```
which is weirdly different.  I suspect it's something to do with https://github.com/rmjarvis/Piff/blob/master/piff/star.py#L233.

**Comment by rmjarvis** [comment]
I think `makeTarget` sets up the WCS such that `wcs.toWorld(image_pos)` returns `field_pos`.  (Assuming the image doesn't have a wcs yet -- you are allowed to provide your own.)  So in that case, I believe the correct center value would be
```
center = image.wcs.toWorld(mom.moments_centroid) - star.field_pos
```
However, I concede the point that these conventions are not all that terribly intuitive, so in you work on this branch, if you feel like something else would make more sense, please feel free to propose a change.

**Comment by rmjarvis** [comment]
I think the general case would be:
```
center = image.wcs.toWorld(mom.moments_centroid) - image.wcs.toWorld(star.image_pos)
```

**Comment by gbernstein** [comment]
Right, that’s the idea.  Though I would consider it permissible to be doing the marginalization with a linearized version of the model with respect to f, x, y (i.e. a Gaussian approx to the likelihood taken as a function of p,f,x,y).  I can’t remember exactly how it was done in this code, but the thought is that if we are willing to iterate these solutions, such an approximation will be good enough to get us to the best fit.

**Comment by rmjarvis** [comment]
Merged via PR #39.
